pkg/api/helpers_test.go

```diff
@@ -397,6 +398,275 @@ func TestMatchTaint(t *testing.T) {
+			},
+			expectTolerated: false,
+		},
+		{
+			description: "toleration with explicit forgiveness can't tolerate taint with no added time, expect not tolerated",
+			toleration: Toleration{
+				Key:                "foo",
+				Operator:           TolerationOpExists,
+				Effect:             TaintEffectNoExecute,
+				ForgivenessSeconds: genForgivenessSeconds(300),
+			},
+			taint: Taint{
+				Key:    "foo",
+				Effect: TaintEffectNoExecute,
+			},
+			expectTolerated: false,
+		},
+	}
+	for _, tc := range testCases {
+		if tolerated := tc.toleration.ToleratesTaint(&tc.taint); tc.expectTolerated != tolerated {
+			t.Errorf("[%s] expect %v, got %v: toleration %+v, taint %s", tc.description, tc.expectTolerated, tolerated, tc.toleration, tc.taint.ToString())
+		}
+	}
+}
+
+func TestTolerationsTolerateTaintsWithFilter(t *testing.T) {
+	testCases := []struct {
+		description       string
+		tolerations       []Toleration
+		taints            []Taint
+		isInterestedTaint taintsFilterFunc
+		expectTolerated   bool
+	}{
+		{
+			description:       "empty tolerations tolerate emtpy taints",
+			tolerations:       []Toleration{},
+			taints:            []Taint{},
+			isInterestedTaint: func(t Taint) bool { return true },
+			expectTolerated:   true,
+		},
+		{
+			description: "non-empty tolerations tolerate emtpy taints",
+			tolerations: []Toleration{
+				{
+					Key:      "foo",
+					Operator: "Exists",
+					Effect:   TaintEffectNoSchedule,
+				},
+			},
+			taints:            []Taint{},
+			isInterestedTaint: func(t Taint) bool { return true },
+			expectTolerated:   true,
+		},
+		{
+			description: "tolerations match all taints, expect tolerated",
+			tolerations: []Toleration{
+				{
+					Key:      "foo",
+					Operator: "Exists",
+					Effect:   TaintEffectNoSchedule,
+				},
+			},
+			taints: []Taint{
+				{
+					Key:    "foo",
+					Effect: TaintEffectNoSchedule,
+				},
+			},
+			isInterestedTaint: func(t Taint) bool { return true },
+			expectTolerated:   true,
+		},
+		{
+			description: "tolerations don't match taints, but no taint is interested, expect tolerated",
+			tolerations: []Toleration{
+				{
+					Key:      "foo",
+					Operator: "Exists",
+					Effect:   TaintEffectNoSchedule,
+				},
+			},
+			taints: []Taint{
+				{
+					Key:    "bar",
+					Effect: TaintEffectNoSchedule,
+				},
+			},
+			isInterestedTaint: func(t Taint) bool { return false },
+			expectTolerated:   true,
+		},
+		{
+			description: "no isInterestedTaint indicated, means all taints are interested, tolerations don't match taints, expect untolerated",
+			tolerations: []Toleration{
+				{
+					Key:      "foo",
+					Operator: "Exists",
+					Effect:   TaintEffectNoSchedule,
+				},
+			},
+			taints: []Taint{
+				{
+					Key:    "bar",
+					Effect: TaintEffectNoSchedule,
+				},
+			},
+			isInterestedTaint: nil,
+			expectTolerated:   false,
+		},
+		{
+			description: "tolerations match interested taints, expect tolerated",
```

Comment:
So basically each node must have its own ``isInterestedTaint`` function. Assuming one pod may tolerate two categories of nodes. E.g. 

* pod with two tolerations: ``region=hpc:NoExecute,Forgiveness=10m``, ``region=slowerhpc:NoExecute,Forgiveness=10m``,
* two categories of nodes ``region=hpc:NoExecute`` and ``region=slowhpc:NoExecute``. 

The pods wants to tolerated taints of both nodes. What are my options? Create two identical pods just with different names (e.g. `pod-hpc`, `pod-slowhpc`) each with its own set of tolerations (``region=hpc:NoExecute,Forgiveness=10m`` for `pod-hpc`, ``region=slowerhpc:NoExecute,Forgiveness=10m`` for `pod-slowhpc`)? If that is the only way, what are expressiveness limitations of taints&toleration language? Something worth describing in docs.

At the same time it does not make much sense to have two following (and alike) taints on the same node:
```
region=A:NoExecute
country=B:NoExecute
```
cause there will never be a pod tolerating both taints (unless the filter function is used).